flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTimeoutTest.java

```diff
@@ -0,0 +1,200 @@
+        // element should be requeued back after timeout
+        tpts.setCurrentTime(110);
+        // deliver initial message after timeout
+        writer.deliverMessage();
+        // flush outstanding mailbox messages containing resubmission of the element
+        writer.flush(false);
+        assertThat(destination).containsExactly(1L, 1L);
+    }
+
+    @Test
+    void writerShouldFailOnTimeoutIfFailOnErrorIsTrue() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, true);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        tpts.setCurrentTime(110);
+        writer.deliverMessage();
+        assertThatExceptionOfType(FlinkRuntimeException.class)
+                .isThrownBy(() -> writer.flush(false))
+                .withCauseInstanceOf(TimeoutException.class)
+                .havingCause()
+                .withMessage("Request timed out after 100ms with failOnTimeout set to true.");
+    }
+
+    @Test
+    void writerShouldDiscardRetriedEntriesOnTimeout() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, false);
+        writer.setShouldFailRequest(true);
+        tpts.setCurrentTime(0);
+        writer.write("1", null);
+        tpts.setCurrentTime(110);
+        // deliver initial message after timeout, message request would fail but elements should be
+        // discarded as timeout already occurred
+        writer.deliverMessage();
+        // flush outstanding mailbox messages to resubmit timed out elements
+        writer.flush(false);
+        assertThat(destination).containsExactly(1L);
+    }
+
+    @Test
+    void writerShouldFailOnFatalError() throws Exception {
+        TestSinkInitContextAnyThreadMailbox context = new TestSinkInitContextAnyThreadMailbox();
+        TestProcessingTimeService tpts = context.getTestProcessingTimeService();
+        TimeoutWriter writer = new TimeoutWriter(context, 1, 10, 100, true);
+        tpts.setCurrentTime(0);
+        writer.setFatalError(new FlinkRuntimeException("Fatal error"));
+        writer.write("1", null);
+        writer.deliverMessage();
+        assertThatExceptionOfType(FlinkRuntimeException.class)
+                .isThrownBy(() -> writer.flush(false))
+                .withMessage("Fatal error");
+    }
+
+    /** Test that the writer can handle a timeout. */
+    private class TimeoutWriter extends AsyncSinkWriter<String, Long> {
+        private Exception fatalError;
+        private final CountDownLatch completionLatch;
+        private Thread submitThread;
+
+        private boolean shouldFailRequest = false;
```

Comment:
Should be ensured thread-safe modification

Reply:
It is only used in scope of tests, we also guarantee no concurrent modifications since it is set by main test thread before latch is triggered to execute the request. I am happy to change but seems to detailed for a mock replacement class